### ISSUE TITLE
Ability to set Keep-Alive timeout response header

### ISSUE THREAD

**Issue by yvespp** [envoyproxy/envoy#17613]
*Description*:
We have set `idleTimeout: 75s` in the envoy lister config and like to set the response header `Keep-Alive: timeout=70` via `response_headers_to_add` so that clients know when to close connections.

Currently the header get's delete by envoy again. I think the relevant code is here: https://github.com/envoyproxy/envoy/blob/cbd8d57e9fc2921d26236db24b6bf91246ba3b4f/source/common/http/conn_manager_utility.cc#L77

*Why?*
We have a lot of Java apps that use the Apache HTTPClient (HTTP/1.1) and don't have a http keep-alive timeout configured and also don't check if connections are still active (default config of the client). This leads to NoHttpResponseExceptions in the apps when evnoy closes the connection after the `idleTimeout`.
If we can set the Keep-Alive header the errors should be reduced as the HTTPClient honors the timeout from that header. See Apache HTTPClient doc: [2.6. Connection keep alive strategy](https://hc.apache.org/httpcomponents-client-4.5.x/current/tutorial/html/connmgmt.html)

We use envoy via Istio but that shouldn't make a difference.

Related to: https://github.com/envoyproxy/envoy/issues/3011
Problems setting Connection headers on cluster responses: https://groups.google.com/g/envoy-users/c/Gop0a7S-vN0/m/eqvvRCKxBgAJ

**Comment by mattklein123** [comment]
This seems like a reasonable optional feature to add. cc @alyssawilk

**Comment by alyssawilk** [comment]
Yeah agreed this is entirely reasonable and you nailed the problem - the router does its header additions then hands the headers to the HCM (https://github.com/envoyproxy/envoy/blob/cbd8d57e9fc2921d26236db24b6bf91246ba3b4f/source/common/http/conn_manager_utility.cc#L363) which does its own sanitization.  We want to clear codec keep-alive but can't differentiate between codec keep-alive and request_headers_to_add keep-alive at the point the sanitization is called.  We should either move the router additions later (tricky) or the hcm sanitization earlier (probably the best bet)

**Comment by sta-szek** [comment]
Hi guys, i faced the same issue, had to stop production rollout.

Just hint from my side, as per mozzilla Keep-Alive header docs: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Keep-Alive
we should also fix the "Connection" header which is also removed conditionally in the line: https://github.com/envoyproxy/envoy/blob/cbd8d57e9fc2921d26236db24b6bf91246ba3b4f/source/common/http/conn_manager_utility.cc#L71 

am i correct?

**Comment by chlunde** [comment]
I wonder if enabling this header as a boolean flag should be supported (which would set it to the configured downstream idle timeout), not just "setting the header including the timeout value"?
This way we could enable "send_keepalive_timeout" by default, without knowing the actual value, and then have other configuration setting the actual idle_timeout depending on the needs of each service.

If we have to set "75 seconds" in two different places, for example in a Istio DestinationRule and a VirtualService or EnvoyFilter,  they can easily get out of sync.

In addition to setting this header, it would be nice if envoy as a HTTP client could parse this header from upstream and close the connection before `min(envoy_internal_idle_timeout, upstream_idle_timeout_from_header  - 1s)`.

HTTP 1.1 Keep-Alive is always prone to race conditions as long as the server closes a connection.

**Comment by l8huang** [comment]
>  We should either move the router additions later (tricky) or the hcm sanitization earlier (probably the best bet)

@alyssawilk if I understand above correctly, with either of them, the Keep-Alive/Connection header added by `response_headers_to_add` will be kept in the response to client, right?

**Comment by alyssawilk** [comment]
possibly (I think so but test!) but if you send them down an HTTP/2 downstream connection it's a protocol violation so be careful with thta.

**Comment by SpecialYang** [comment]
I met the same issue. The policy response_headers_to_add or response_headers_to_set can't work.

@l8huang @alyssawilk
